src/qs8-gemm/MRx4c16-wasmusdot.c.in

```diff
@@ -0,0 +1,230 @@
+$PARAMS_UNION = {"QC8": "xnn_qs8_qc8w_conv_minmax_params", "QD8": "xnn_f32_minmax_params", "QS8": "xnn_qs8_conv_minmax_params"}[DATATYPE]
+$OUT_T = "float" if DATATYPE == "QD8" else "int8_t"
+void xnn_${DATATYPE_SPEC}_gemm_minmax${REQUANTIZATION_SPEC}_ukernel_${MR}x4c16__wasmusdot(
+    size_t mr,
+    size_t nc,
+    size_t kc,
+    const int8_t* restrict a,
+    size_t a_stride,
+    const void* restrict w,
+    ${OUT_T}* restrict c,
+    size_t cm_stride,
+    size_t cn_stride,
+    $if DATATYPE == "QD8":
+      const union ${PARAMS_UNION} params[restrict XNN_MIN_ELEMENTS(1)],
+      const struct xnn_qd8_quantization_params quantization_params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
+    $else:
+      const union ${PARAMS_UNION} params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
+{
+  assert(mr != 0);
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
+  assert(kc % sizeof(int8_t) == 0);
+  assert(a != NULL);
+  assert(w != NULL);
+  assert(c != NULL);
+
+  kc = round_up_po2(kc, 16 * sizeof(int8_t));
+  const int8_t* a0 = a;
+  ${OUT_T}* c0 = c;
+  $for M in range(1, MR):
+    const int8_t* a${M} = (const int8_t*) ((uintptr_t) a${M-1} + a_stride);
+    ${OUT_T}* c${M} = (${OUT_T}*) ((uintptr_t) c${M-1} + cm_stride);
+    $if M % 2 == 0:
+      if XNN_UNPREDICTABLE(mr <= ${M}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+    $elif M + 1 == MR:
+      if XNN_UNPREDICTABLE(mr != ${M+1}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+    $else:
+      if XNN_UNPREDICTABLE(mr < ${M+1}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+
+  const v128_t vsign_mask = wasm_u8x16_const_splat(UINT8_C(0x80));
+  do {
+    $if DATATYPE == "QD8":
+      v128_t vksum0 = wasm_v128_load32_zero(w);
+      $for N in range(1, 4):
+        v128_t vksum${N} = wasm_v128_load32_zero((const int32_t*) w + ${N});
+      $for M in range(MR):
+        const v128_t vinput_zero_point${M} = wasm_v128_load32_splat(&quantization_params[${M}].zero_point);
+      $for M in range(MR):
+        $for N in range(4):
+          v128_t vacc${M}x${N} = wasm_i32x4_mul(vksum${N}, vinput_zero_point${M});
+    $else:
+      v128_t vacc0x0 = wasm_v128_load32_zero(w);
+      $for N in range(1, 4):
+        v128_t vacc0x${N} = wasm_v128_load32_zero((const int32_t*) w + ${N});
+      $for M in range(1, MR):
+        $for N in range(4):
+          v128_t vacc${M}x${N} = vacc0x${N};
+    w = (const int32_t*) w + 4;
+
+    size_t k = kc;
+    do {
+      $for M in range(MR):
+        const v128_t va${M} = wasm_v128_xor(wasm_v128_load(a${M}), vsign_mask);
+        a${M} += 16;
+
+      $for N in range(4):
+        $if N == 0:
+          const v128_t vb${N} = wasm_v128_load(w);
+        $else:
+          const v128_t vb${N} = wasm_v128_load((const int8_t*) w + ${N * 16});
+
+        $for M in range(MR):
+          vacc${M}x${N} = wasm_i32x4_relaxed_dot_i8x16_i7x16_add(vb${N}, va${M}, vacc${M}x${N});
```

Comment:
If weights are signed int8 this will overflow if the implementation is vpmaddubsw?

Reply:
The implementation of vpmaddubsw will not pass the check CheckWAsmUSDOT as it has a different saturation behavior.